Modernize Project and Add Support for FIPS Systems

mcdonnnj opened the pull request:
# <!--- Provide a general summary of your changes in the Title above -->

## 🗣 Description
This PR updates this project to transition from TravisCI to GitHub Actions, adds an appropriate Lineage configuration file, and adds support for FIPS systems to resolve #9 . There are also some updates to code logic to make it more easily extensible for new hash types and to DRY out some code.
<!--- Describe your changes in detail -->

## 💭 Motivation and Context
@bra1ncramp pointed out that on a systems in FIPS mode the project as it is will not correctly run. As part of implementing a fix for that use case this project needed to be updated from its skeleton and modernized to align with how we do projects now.
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

## 🧪 Testing
Automated tests work as expected.
Edit:
I have tested this code on a FIPS system both installed as a library, and in standalone mode with Ansible as described in the README. There were no issues and I confirmed that it was successful in creating hashing objects with the `usedforsecurity` keyword argument.
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## 📷 Screenshots (if appropriate)

## 🚥 Types of Changes

<!--- What types of changes does your code introduce? -->
<!--- Put an `x` in all the boxes that apply: -->

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (causes existing functionality to change)

## ✅ Checklist

<!--- Go over all the following points, and put an `x` in all the
boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask.
We're here to help! -->

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

mcdonnnj committed:
Merge branch 'develop' of github.com:cisagov/skeleton-python-library into improvement/add_fips_compatibility

mcdonnnj committed:
Fix Lineage configuration and update isort configuration.

mcdonnnj committed:
Fixes from isort and mypy

mcdonnnj committed:
Make it easier to add new hash algorithms.

Adding a new hashing algorithm now consists of adding the hashlib constructor
name in DESIRED_HASHES and adding a new regex to match output for the desired
algorithm to HASH_REGEXES.
This also resulted in DRYing out code used for getting indicators and hashing
a file.

mcdonnnj committed:
Add support for systems in FIPS mode

When a system is in FIPS mode not all hashing algorithms are usable by default.
The most notable of these is MD5. When trying to generate an MD5 hash in FIPS
mode a ValueError exception is raised. Support was added to get around this for
uses such as this project by passing a "usedforsecurity=False" keyword
argument to the constructor. However, this is not universallty supported and on
some systems this constructor form will result in a TypeError as a result of an
unsuported keyword argument. I observed this on both my MacBook and on an
Ubuntu WSL instance when testing. A try/except block allows us to support both
configurations (hopefully).

mcdonnnj committed:
Fix typo in function comment

mcdonnnj committed:
Code cleanup from pylint feedback

Replace accidental remap of built-in function hash.
Change calls to the logging library to use the preferred printf style format
for outputs.

mcdonnnj committed:
Fix __init__ setup to appease mypy and LGTM analysis

Using __all__ = [""] made mypy happy, but LGTM analysis was very not happy.

jsf9k reviewed (approved):
These changes look good to me.  Have they been tested against a FIPS system?

hillaryj commented on src/ioc_scan/ioc_scanner.py:
```diff
@@ -9,6 +9,7 @@
 This script should be run as a priveledged user.
```
Nit: should be `privileged`

mcdonnnj replied:
Good catch! Fixed in 0f54aa1

hillaryj commented on tests/test_ioc_scan.py:
```diff
@@ -1,17 +1,18 @@
 import pytest
 
+# cisagov Libraries
 import ioc_scan
```
Is it necessary to import both `ioc_scan` and specific components of it?

mcdonnnj replied:
The `ioc_scan` package doesn't export those component so the `import ioc_scan` is just to get `ioc_scan.__version__`. We need to import the specific components to access them for testing.

hillaryj replied:
Roger that!

hillaryj reviewed (approved):
👍  Left some minor comments

mcdonnnj committed:
Fix misspelling in ioc_scanner.py description comment

hillaryj reviewed (approved):
👍

dav3r commented on .github/CODEOWNERS:
```diff
@@ -0,0 +1,7 @@
+# the repo. Unless a later match takes precedence,
+# these owners will be requested for review when someone
+# opens a pull request.
+*       @dav3r @felddy @hillaryj @jsf9k @mcdonnnj @cisagov/team-ois
```
@mcdonnnj - please remove @hillaryj and @jsf9k as code owners, as part of our effort to be more targeted with owners.

mcdonnnj replied:
Should we also remove `@cisagov/team-ois`?

dav3r replied:
Yes, please remove `@cisagov/team-ois` too - thanks!

mcdonnnj replied:
👍 Trimmed in 4480be7

dav3r commented on bump_version.sh:
```diff
@@ -0,0 +1,51 @@
+set -o errexit
+set -o pipefail
+
+VERSION_FILE=src/example/_version.py
```
`example` --> `ioc_scan`

mcdonnnj replied:
Ah, good catch. Fixed in 48a4862

dav3r commented on src/ioc_scan/ioc_scanner.py:
```diff
@@ -48,30 +49,51 @@
+    return tuple(getattr(hashlib, algo) for algo in available_hashers)
+
+
+# List of hash algorithms we want to use on files. These shoudl correspond to
```
Typo: `shoudl` --> `should`

mcdonnnj replied:
Oops. Fixed in 64f0e9d

dav3r reviewed (approved):
This is looking solid @mcdonnnj - I just noted a few minor thangs.  🦘

mcdonnnj committed:
Fix typo shoudl -> should

mcdonnnj committed:
Fix location of _version.py in bump_version.sh

mcdonnnj committed:
Trim members in CODEOWNERS

mcdonnnj commented:
> 
> 
> These changes look good to me. Have they been tested against a FIPS system?

@jsf9k I'm waiting for access to the test server (which is a FIPS system). Once I have confirmed functionality, I will merge unless there are outstanding change requests.

mcdonnnj committed:
Fix missed variable replacement and add additional test assert

Missed one instance of hash->item which caused hits to fail to tally. Added an
additional assert to test_scan_stdin to ensure the tally result is what we
expected.

mcdonnnj committed:
Bump version from 1.1.0 to 1.2.0
